refactor(compiler-cli): deprecate the `fullTemplateTypeCheck` compile…

JoostK opened the pull request:
…r option

When compiling your application using the AOT compiler, your templates
are type-checked according to a certain strictness level. Before Angular 9
there existed only two strictness levels of template type checking as
determined by [the `fullTemplateTypeCheck` compiler option](guide/angular-compiler-options).
In version 9 the `strictTemplates` family of compiler options has been
introduced as a more fine-grained approach to configuring how strict your
templates are being type-checked.

The `fullTemplateTypeCheck` flag is being deprecated in favor of the new
`strictTemplates` option and its related compiler options. Projects that
currently have `fullTemplateTypeCheck: true` configured can migrate to
the following set of compiler options to achieve the same level of
type-checking.

```json
{
  "angularCompilerOptions": {
    "strictTemplates": true,
    "strictInputTypes": false,
    "strictNullInputTypes": false,
    "strictAttributeTypes": false,
    "strictOutputEventTypes": false,
    "strictDomEventTypes": false,
    "strictDomLocalRefTypes": false,
    "strictSafeNavigationTypes": false,
    "strictContextGenerics": false,
  }
}
```

JoostK committed:
refactor(compiler-cli): deprecate the `fullTemplateTypeCheck` compiler option

When compiling your application using the AOT compiler, your templates
are type-checked according to a certain strictness level. Before Angular 9
there existed only two strictness levels of template type checking as
determined by [the `fullTemplateTypeCheck` compiler option](guide/angular-compiler-options).
In version 9 the `strictTemplates` family of compiler options has been
introduced as a more fine-grained approach to configuring how strict your
templates are being type-checked.

The `fullTemplateTypeCheck` flag is being deprecated in favor of the new
`strictTemplates` option and its related compiler options. Projects that
currently have `fullTemplateTypeCheck: true` configured can migrate to
the following set of compiler options to achieve the same level of
type-checking.

```json
{
  "angularCompilerOptions": {
    "strictTemplates": true,
    "strictInputTypes": false,
    "strictNullInputTypes": false,
    "strictAttributeTypes": false,
    "strictOutputEventTypes": false,
    "strictDomEventTypes": false,
    "strictDomLocalRefTypes": false,
    "strictSafeNavigationTypes": false,
    "strictContextGenerics": false,
  }
}
```

mary-poppins commented:
You can preview 6ed8ed4 at https://pr43224-6ed8ed4.ngbuilds.io/.

atscott reviewed (changes requested):
I think you'll also want to update the actual readme for the compiler options: https://github.com/angular/angular/blob/master/aio/content/guide/angular-compiler-options.md


* Note there that it's deperated
* Don't use it in examples (use `strictTemplates` instead)

atscott commented on aio/content/guide/deprecations.md:
```diff
@@ -443,6 +444,22 @@ If you rely on the behavior that the same object instance should cause change de
+templates are being type-checked.
+
+The `fullTemplateTypeCheck` flag is being deprecated in favor of the new `strictTemplates` option and its related compiler options.
+When `fullTemplateTypeCheck` is eventually removed, its default value of `false` will effectively change to `true` which results in type-checking
```
Have we decided exactly what we will do when `fullTemplateTypeCheck` is removed? 

Maybe we shouldn't guess at what _might_ happen but instead just point people to `strictTemplates` as the replacement. We'll have to have a separate breaking changes note when we end up removing `fullTemplateTypeCheck` anyways.

JoostK replied:
The primary goal of the deprecation is to get rid of the "basic mode" that is currently available using `fullTemplateTypeCheck: false`, which is the default currently. So from that perspective the eventual removal has to mean that the option is treated as if it were `true` al along, which is indeed a breaking change for projects for which either 1) the `fullTemplateTypeCheck` option has not been configured and `strictTemplates` is not `true`, or 2) the option is explicitly disabled.

The tricky bit about pointing people to `strictTemplates` is that it can only get you even more strict template type-checking. The breaking change only concerns projects that do not have `fullTemplateTypeCheck` at all, in which case the `strictTemplates` option won't help in suppressing template type-errors that may be caused by the breaking change.

Concretely, the following internal type-check flags are to be removed, as they can only be `true` if `fullTemplateTypeCheck` is removed:

- `alwaysCheckSchemaInTemplateBodies`
- `checkTemplateBodies`
- `checkTypeOfNonDomReferences`
- `checkTypeOfPipes`

We discussed in a FW meeting a while ago (when discussing the deprecation of `fullTemplateTypeCheck`) that we don't intent to add new strictness flags to allow disabling some of the above options. Hence, this paragraph goes into detail which template features may start reporting errors once `fullTemplateTypeCheck` is eventually removed.

JoostK replied:
I realized that maybe we could/should push for enabling `fullTemplateTypeCheck` as a separate breaking change before removing the option entirely (perhaps even in v13). Projects that don't have the option configured will then become aware of any new errors and have the choice to temporarily disable `fullTemplateTypeCheck`, at which point they explicitly opt in to the deprecated behavior.

JoostK committed:
docs: remove legacy compiler options in library schematics example

The `skipTemplateCodegen`, `strictMetadataEmit` and
`enableResourceInlining` options are specific to ViewEngine and have no
effect when using the Ivy compiler. Since libraries can no longer be
compiled using ViewEngine from Angular 13, this commit removes these
options entirely.

The `annotateForClosureCompiler` is also no longer recommended to use,
as Closure Compiler is not supported outside of Bazel.

Finally, the `fullTemplateTypeCheck` flag is changed into
`strictTemplates` as the former flag is being deprecated. Enabling
`strictTemplates` does result in more strict type-checking than with
only `fullTemplateTypeCheck` enabled, but simply removing
`fullTemplateTypeCheck` right away would result in less strict
type-checking which is not desired.

mary-poppins commented:
You can preview 7273fb0 at https://pr43224-7273fb0.ngbuilds.io/.

JoostK commented:
@atscott good points on the examples and Angular compiler option docs page, I have added a fixup commit to address that.

atscott reviewed (approved):
LGTM, thanks for the updates!

mary-poppins commented:
You can preview c4b481b at https://pr43224-c4b481b.ngbuilds.io/.

JoostK commented on aio/content/examples/schematics-for-libraries/projects/my-lib/tsconfig.lib.json:
```diff
@@ -17,12 +17,8 @@
-    "fullTemplateTypeCheck": true,
-    "strictInjectionParameters": true,
-    "enableResourceInlining": true
+    "strictTemplates": true,
```
Can you point me to where those are defined?

JoostK replied:
I have been told that these options are handled by StackBlitz; we'll reach out to get them updated. Thanks for letting us know about this.

mary-poppins commented:
You can preview b1dba8e at https://pr43224-b1dba8e.ngbuilds.io/.

atscott reviewed (approved):
🥳

alxhub reviewed (approved):
Reviewed-for: public-api

IgorMinar reviewed (approved):
LGTM! Thank you!

Reviewed-for: global-docs-approvers, docs-packaging-and-releasing

iRealNirmal commented on aio/content/guide/deprecations.md:
```diff
@@ -525,6 +526,36 @@ class SubmitButton {
+{
+  "angularCompilerOptions": {
+    ...
+    "strictTemplates": true,
```
if PR [https://github.com/angular/angular/pull/43545](https://github.com/angular/angular/pull/43545) gets merged before this you can update this snippet as well.  

else It has to be taken care in the PR [https://github.com/angular/angular/pull/43545](https://github.com/angular/angular/pull/43545).

atscott reviewed (approved):
reviewed-for: public-api

mary-poppins commented:
You can preview d7ab46c at https://pr43224-d7ab46c.ngbuilds.io/.

jessicajaniuk reviewed (approved):
LGTM :cookie:

reviewed-for: public-api

dylhunn commented:
This PR was merged into the repository by commit cc96f322df85f5159ff51c590518acf14f242264.
